jax/_src/scipy/special.py

```diff
@@ -657,3 +659,92 @@ def i1e(x):
 def i1(x):
   x, = _promote_args_inexact("i1", x)
   return lax.mul(lax.exp(lax.abs(x)), lax.bessel_i1e(x))
+
+def lpmn(m, n, z):
+  r"""The normalized associated Legendre functions of the first kind.
+
+  In the descriptions of the associated Legendre functions (ALFs), we use `l`
+  to denote the degree (`l_max` is the maximum degree) and `m` to denote
+  the order. The ALFs of the first kind are often used in spherical harmonics.
+  The spherical harmonic of degree `l` and order `m` can be written as
+  .. math::
+    \begin{equation}
+      Y_l^m (\theta, \phi) = N_l^m P_l^m(\cos \theta) e^{i m \phi},
+    \end{equation}
+  where
+  .. math::
+    \begin{equation}
+      N_l^m = \sqrt{\displaystyle \frac{\left(2l + 1\right)
+              \left(l-m\right)!}{4 \pi \left(l+m\right)!}}
+    \end{equation}
+  represents the normalization constant, and θ and φ are the colatitude and
+  longitude, respectively. :math:`N_l^m` is chosen in the way that the spherical
+  harmonics form a set of orthonormal basis function of :math:`L^2(S^2)`.
+  The normalized associated Legendre functions can be expressed as
+  .. math::
+    \begin{equation}
+      \bar{P}_l^m = N_l^m P_l^m.
+    \end{equation}
+
+  Args:
+    m: The maximum order of the associated Legendre functions.
+    n: The maximum degree of the associated Legendre function, often called
+      `l` in describing ALFs. Both the degrees and orders are
+      `[0, 1, 2, ..., l_max]`.
+    z: A vector of type `float32` and `float64` containing the sampling
+      points in spherical coordinates, at which the ALFs are computed;
+      and `z` is essentially `cos(θ)`.
+
+  Returns:
+    The 3D array of shape `(l_max + 1, l_max + 1, len(z))` containing the
+    normalized values of the ALFs at `z`. The dimensions follow the order
+    of order `m`, degree `l`, and `z`.
+
+  Raises:
+    TypeError if type of `z` is not in (float32, float64).
+    ValueError if `z` is not 1D.
+    NotImplementedError if `m!=n`.
+  """
+  dtype = lax.dtype(z)
+  if dtype not in (jnp.float32, jnp.float64):
+    raise TypeError(
+        'x.dtype={} is not supported, see docstring for supported types.'
+        .format(dtype))
+
+  if z.ndim != 1:
+    raise ValueError('x must be a 1D array.')
+
```

Comment:
Before comparing `m != n`, we should call `core.concrete_or_error()` on each value (it leads to better error messages in the case that `m` or `n` are traced).

Reply:
Done. Could you explain the concept of trace? Or any reference/document will help me understand. Thanks!

Reply:
This doc might be useful: https://jax.readthedocs.io/en/latest/notebooks/thinking_in_jax.html